### DISCUSSION TITLE
Slow (and weak) AutoCompletion

### DISCUSSION THREAD

**Discussion by Georg97** [helix-editor/helix#6899] [General]
# Issue
Compared to my NeoVim config with lsp-zero and cmp, the helix AutoCompletion just seems to lack a lot to be a serious competitor.

My problems are:

- needs about 0.5 - 1 second unit autocomplete suggestions are shown (Maybe I can configure this somehow)
- For some specific languages the AutoCompletion just does not show every possibilty (e.g. java with jdtls and scss with vscode-language-server)

# My Setup

I installed `vscode-language-server` through npm and jdtls by extracting to `~/.local/bin/jdtls-files` and creating a symlink `~/.local/bin/jdtls -> ./jdtls-files/bin/jdtls`. 

`hx --health` has following output for java:

![image](https://user-images.githubusercontent.com/75206704/235003732-6c57a4fb-7e47-437f-bb7e-aa00217a9f6b.png)

posting my whole NeoVim config would be a bit too much I guess. It can be found [here](https://github.com/Georg97/nvimrc)

# Screenshots
![image](https://user-images.githubusercontent.com/75206704/235004157-90e34514-1d53-41a4-a849-877863d731d4.png)
auto complete with neovim

![image](https://user-images.githubusercontent.com/75206704/235004311-20c3c2e6-5601-4ce1-a7e7-95544235c4f8.png)
autocomplete with helix. It shows less results and `addClassNames` ist not even existent

# misc
The main issue is the missing completion. I can live with the slower results, but would nice to address them somehow.

I have similar problems with css/scss files, but I think this one example shoud be enough for now. Maybe when I find a solution for this, the other one will also vanish.

**When coding typescript I don't have this issue. It's just slow, but shows all possibilites for code completion**

I don't know if this is a bug or might be feature request. I guess there is also a high chance, that I have just configured jdtls wrong. I just can't find any practical examples on how to configure helix for java development. Any help would be much appreciated!

**Comment by clo4** [comment]
Not sure about the second issue, but for the first one, add this to your helix config (use `:config-open`)
```toml
[editor]
idle-timeout = 0
completion-trigger-len = 1
```

**Reply by Georg97** [reply]
Great, thanks for that one. That was what I was hoping for.

**Comment by kirawi** [comment]
Helix shows all the auto-completion results the LSP sends, so there isn't anything that can be done client-side about it.

**Reply by Georg97** [reply]
Yes I do understand that, technically... But how does neovim do that then? I'm just curious

**Reply by pascalkuthe** [reply]
It seems like you might still be using an older helix version (22.12 instead of 22.3) which didn't include snippet support. Many LSPs will just send less completions when the client disables snippets

**Reply by Georg97** [reply]
`hx --version` says I run version 23.03. And the missing auto completions are not snippets, they are missing member functions and properties
![image](https://user-images.githubusercontent.com/75206704/235481126-3dafd2cf-a9f1-4b55-b442-7d042f6d4f6c.png)

**Reply by clo4** [reply]
Is it possible that your LS is (maybe incorrectly, maybe not) not seeing a capability being reported, and as a result sending fewer completions? I don't know anything about your LS or what capabilities Helix reports, so that's a total shot in the dark from me
